### ISSUE TITLE
`GatherElements op: Out of range value in index tensor` — YOLO26x with CoreML/ONNX on Apple Silicon

### ISSUE THREAD

**Issue by KingTimer12** [ultralytics/inference#148]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/PROJECT_NAME/issues) and found no similar bug report.

### Ultralytics Component

Setup

### Bug

```
Registering 1 execution providers (primary: CoreMLExecutionProvider)
2026-05-08 10:04:21.609 vision-client[92285:938854] 2026-05-08 10:04:21.609732 [E:onnxruntime:, sequential_executor.cc:572 ExecuteKernel] Non-zero status code returned while running GatherElements node. Name:'/model.23/GatherElements_1' Status Message: /Users/runner/work/ort-artifacts/ort-artifacts/onnxruntime/onnxruntime/core/providers/cpu/tensor/gather_elements.cc:150 void onnxruntime::core_impl(const Tensor *, const Tensor *, Tensor *, int64_t, concurrency::ThreadPool *) [Tin = long long] GatherElements op: Out of range value in index tensor
Registering 1 execution providers (primary: CoreMLExecutionProvider)
2026-05-08 10:04:22.209 vision-client[92285:938854] 2026-05-08 10:04:22.209716 [E:onnxruntime:, sequential_executor.cc:572 ExecuteKernel] Non-zero status code returned while running GatherElements node. Name:'/model.23/GatherElements_1' Status Message: /Users/runner/work/ort-artifacts/ort-artifacts/onnxruntime/onnxruntime/core/providers/cpu/tensor/gather_elements.cc:150 void onnxruntime::core_impl(const Tensor *, const Tensor *, Tensor *, int64_t, concurrency::ThreadPool *) [Tin = long long] GatherElements op: Out of range value in index tensor
```

### Environment

| Key | Value |
|-----|-------|
| Platform | macOS (Apple M4) |
| Runtime | ONNX Runtime + CoreMLExecutionProvider |
| Model | YOLO26x (`best.onnx`) |
| Ultralytics | 8.4.47 |
| Python | 3.12.12 |
| PyTorch | 2.11.0 |

### Steps
1. Export a YOLO26x model to ONNX
2. Load it in an app using ORT with `CoreMLExecutionProvider` on Apple Silicon
3. Run inference on any frame/image

### Code
```python
# Export python area
from ultralytics import YOLO
model = YOLO("models/best.pt")
model.export(format="onnx", imgsz=640, half=True, simplify=True, opset=12, nms=True)
```
```rust
// Inference rust area
pub fn load(model_path: &str, confidence: f32, iou: f32, _imgsz: usize) -> anyhow::Result<Self> {
        let device = if cfg!(target_os = "macos") {
            Device::Mps
        } else if cfg!(target_os = "windows") {
            Device::Cuda(0)
        } else {
            Device::Cpu
        };
        let config = InferenceConfig::new()
            .with_confidence(confidence)
            .with_iou(iou)
            .with_device(device)
            .with_save(false);

        let mut model = YOLOModel::load_with_config(model_path, config)
            .map_err(|e| anyhow::anyhow!("Model load failed: {e}"))?;

        if let Err(e) = model.warmup() {
            let msg = e.to_string();
            if !msg.contains("GatherElements") && !msg.contains("Out of range") {
                return Err(anyhow::anyhow!("Warmup failed: {msg}"));
            }
        }
        eprintln!("[detector] Loaded: {} provider={}", model_path, model.execution_provider());
        Ok(Self { model })
    }
```

### Additional

Is there a known export flag or workaround to prevent `model.23` from generating out-of-range indices under CoreMLExecutionProvider? Is this a known limitation of YOLO26x with ORT + CoreML?

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @KingTimer12, thank you for submitting a `ultralytics/inference` 🚀 issue. This is an automated response to help us triage your report quickly, and an Ultralytics engineer will also assist soon 😊

To help us investigate this `GatherElements` runtime issue efficiently, please ensure the following details are included in this thread:

1. For bug reports:
   - A clear and concise description of the bug
   - A minimum reproducible example [MRE](https://docs.ultralytics.com/help/minimum-reproducible-example/) that demonstrates the issue
   - Your environment details (OS, Python version, package versions)
   - Expected behavior vs. actual behavior
   - Any error messages or logs related to the issue

2. For feature requests:
   - A clear and concise description of the proposed feature
   - The problem this feature would solve
   - Any alternative solutions you've considered

3. For questions:
   - Provide as much context as possible about your question
   - Include any research you've already done on the topic
   - Specify which parts of the [documentation](https://docs.ultralytics.com/), if any, you've already consulted

From your report, it looks like you’ve already included helpful environment details, logs, and export/inference snippets 🙌 If possible, please also add a fully self-contained [MRE](https://docs.ultralytics.com/help/minimum-reproducible-example/) that another maintainer can run end-to-end without modification, including:
- The exact model export command used
- Whether the issue reproduces with the same ONNX file outside your app flow
- Any preprocessing details that happen before inference
- The exact `ultralytics/inference` version or commit being used
- Whether the issue occurs consistently on every image or only specific inputs

Please make sure you've searched existing issues to avoid duplicates. If you need to add any additional information, please comment on this issue. 🔍

Thank you for your contribution to improving our project!

**Comment by onuralpszr** [comment]
@KingTimer12 hello let me check it out quickly

**Comment by onuralpszr** [comment]
Thanks for the report! I reproduced this locally and traced it to two separate
bugs in how ORT rc.12's CoreML execution provider handles ONNX models.
Both are now fixed details below.

## What was happening

There were actually **two distinct errors** depending on the model's precision:

| Scenario | Error | When |
|---|---|---|
| Any ONNX model on `--device mps` | `GatherElements op: Out of range value in index tensor` | model load / warmup only |
| FP16 ONNX + CoreML | `Feature images is required but not specified` | every inference call |

### Error 1 GatherElements during warmup

When the model loads it runs a single "warmup" pass with an all-zeros
`(1, 3, 640, 640)` tensor to pre-allocate memory and trigger JIT compilation.

The YOLO DFL decode head (`model.23`) uses a `GatherElements` op to look up
distribution values from a learned table. With an all-zeros input, every
logit is zero, the softmax output is uniform, and the argmax produces
edge-case indices that fall outside the table bounds.

**This is purely a warmup artefact.** Real images produce a natural spread of
logits that result in valid indices  actual inference always works correctly.

### Error 2  FP16 + CoreML NeuralNetwork miscompilation (ORT rc.12 bug)

When ORT rc.12 compiles a FP16 ONNX graph into a CoreML **NeuralNetwork**
model (Core ML ≤ 4 format), it mishandles the FP32 → FP16 cast chain at the
model input. The resulting `.mlpackage` ends up with an unresolvable feature
descriptor, causing every inference call to fail not just warmup.

This is specific to ORT rc.12 (~1.24.x). Python ORT 1.25.1 handles the same
FP16 graph without any error. Once ort update their side maybe I can change to simpler for now I will keep eye on that problem

## Fixes

### Fix 1  Switch CoreML EP to MLProgram format

**MLProgram** (Core ML 5, macOS 12+) natively supports FP16 ops and compiles
the ONNX graph correctly under ORT rc.12. Switching away from the legacy
NeuralNetwork format eliminates Error 2 entirely.

The CoreML compilation cache subdirectory now includes a format suffix
(`_mlprogram` / `_neuralnet`) so a format change causes a cache miss instead
of silently reusing a stale `.mlpackage`.

### Fix 2 Graceful warmup error handling

Error 1 is silently in `warmup()` with a narrow pattern match
(`GatherElements` + `Out of range`). A warning is logged and warmup continues.
Any other error still propagates normally  the filter is intentionally tight.

### Fix 3 Runtime macOS version fallback

MLProgram requires macOS 12 or newer. On older hosts (macOS 11 / Intel Macs
that haven't been upgraded), the library now detects the OS version at runtime
and automatically falls back to NeuralNetwork format with an explanatory
warning. FP32 models work correctly in that path; FP16 models will warn that
they may fail and suggest re-exporting with `half=False`.

## Export matrix all variants now work

Tested all combinations of opset × precision against `bus.jpg`: (our assets) - Apple Silicon M2 machine so speed might vary in your machine of course 

| Export config | CPU | MPS (CoreML) | Detections |
|---|---:|---:|---|
| op12 fp32 | 50.8 ms | 47.3 ms | 4 persons, 1 bus ✓ |
| op12 fp16 | 63.2 ms | 32.9 ms | 4 persons, 1 bus ✓ |
| op17 fp32 | 51.1 ms | 32.1 ms | 4 persons, 1 bus ✓ |
| **op17 fp16** | 59.9 ms | **25.1 ms** | 4 persons, 1 bus ✓ |
| op19 fp32 | 50.5 ms | 28.8 ms | 4 persons, 1 bus ✓ |
| op19 fp16 | 59.6 ms | 27.5 ms | 4 persons, 1 bus ✓ |

**Opset version has no effect on correctness** the GatherElements crash is a runtime input issue, not an export artefact. Speed Fp16 is natural slow because GPU/MPS use FP16 so for test case I tested just in case for error and crashes

## Compatibility

| macOS | FP32 ONNX | FP16 ONNX |
|---|---|---|
| 12+ (Monterey and newer) | ✓ | ✓ |
| 11 (Big Sur) | ✓ | ⚠️ re-export with `half=False` |
| 10.x | ✓ | ⚠️ re-export with `half=False` |

All Apple Silicon Macs shipped with macOS 12+, so the warning path is primarily relevant for older Intel Macs.

**Fix is in:** https://github.com/ultralytics/inference/pull/149

**Comment by onuralpszr** [comment]
@KingTimer12 I posted my finding and PR link so if you can test that would be awesome.

**Comment by KingTimer12** [comment]
ill, one sec

**Comment by onuralpszr** [comment]
> ill, one sec

Great to hear it

**Comment by KingTimer12** [comment]
its working perfectly!! Great job!! Thank u

**Comment by onuralpszr** [comment]
> its working perfectly!! Great job!! Thank u

You are welcome and new version shipped, it should be release soon so you can upgrade version and use it :)

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has now been merged in [PR #149: `ultralytics-inference 0.0.13` fix: resolve CoreML FP16 crash and warmup GatherElements error](https://github.com/ultralytics/inference/pull/149) by @onuralpszr.

The merged changes that may resolve this include:

- updating `ultralytics-inference` to `0.0.13`
- improving CoreML behavior on macOS by selecting a safer model format based on macOS version
- using `MLProgram` on macOS 12+ and falling back to `NeuralNetwork` on older or unknown macOS versions
- adding warnings for older macOS environments that may have FP16 CoreML compatibility issues
- isolating CoreML cache directories by model format to avoid cache collisions
- making warmup more fault-tolerant by:
  - capturing warmup inference errors instead of failing immediately
  - ignoring the known benign CoreML `GatherElements` out-of-range error triggered by all-zero dummy warmup input
  - still surfacing real inference errors normally

These changes should help reduce false startup failures and improve CoreML reliability, especially on Apple devices and older macOS setups.

Credit to @onuralpszr for the fix, and thanks as well to everyone who helped report, investigate, and narrow this down.

If you'd like to test the fix, you can try either:

- the released version:
  ```bash
  pip install -U ultralytics-inference>=0.0.13
  ```
- or the latest `main` branch directly:
  ```bash
  pip install git+https://github.com/ultralytics/inference.git@main
  ```

If you prefer, you can also wait for the next official PyPI release in your normal update cycle.

Please let us know whether these changes resolve the issue on your setup. Thanks 🙏 for reporting this, and any additional feedback is welcome if the issue still persists.
